### ISSUE TITLE
A higher-level LocalAPI specifically for GUI-centred events and operations

### ISSUE THREAD

**Issue by dblohm7** [tailscale/tailscale#13392]
TL;DR: Can we move a bunch of implemented-per-platform notification processing out of the GUIs and into `tailscaled` itself?

In addition to the other LocalAPI changes described in https://github.com/tailscale/tailscale/issues/13390:

Each platform's GUI has its own separate implementation of code to extract useful information out of LocalAPI notifications and digest them into relevant events that the GUIs must act upon -- the clients pods have mused about sharing as much of this code as possible within the `tailscaled` LocalAPI surface itself. While moving to notifications with incremental updates is undoubtedly a huge part of this, the clients pods have discussed other, additional enhancements that would be helpful to the GUIs:

* The notifications delivered via `WatchIPNBus` do not currently include additional information that could be relevant. For example:
  - When a netmap is received, the GUI needs to compute whether network lock is currently enabled or whether it has been disabled (based on comparing against state from the previous netmap), and if so the GUI then needs to perform a synchronous query to determine the actual network lock status. All of this could be sent to the GUI from `tailscaled` within a single notification.
  -  Every time the GUI receives a new netmap, it needs to determine whether there are location-based exit nodes, and if so make a synchronous LocalAPI call to retrieve the suggested node. This information could be delivered more cleanly in its own type of notification, instead of requiring each GUI to do its own digging.

Another issue we deal with on the GUI side is that we have a larger domain of states than those provided via `ipn.State`. In the Windows GUI, we essentially have a private `uiState` enumeration that operates as a superset of `ipn.State`. Many of these `uiState` values are resolved via munging of the data we receive via IPN bus notifications, sometimes requiring additional API round-trips. Can we add some stuff to the LocalAPI surface that enables `tailscaled` itself to be more aware of these higher-level states?

Obviously there would be many design details to work out: do all platforms share a common set of GUI states, or do we provide some kind of mapping mechanism that GUIs could use to declare how they want state to be perceived and events to be delivered? Something else?

**Comment by nickkhyl** [comment]
One important factor to keep in mind is that more than one LocalAPI client might be connected at the same time. On Windows, this includes multiple logged-in users, but it's also relevant to all platforms that support the CLI and/or web client.
